compiler/src/iree/compiler/Dialect/Util/Transforms/ConvertPrimitiveType.cpp

```diff
@@ -210,11 +210,108 @@ struct ConvertTypeSensitiveArithCastOp : public OpConversionPattern<OpTy> {
   }
 };
 
+// Truncates input operand if it is a constant and is out of range of target
+// type.
+Value truncateOutOfRangeValue(Value operand,
+                              ConversionPatternRewriter &rewriter,
+                              TypeConverter *typeConverter) {
+  if (!operand.getDefiningOp()) {
+    return operand;
+  }
+  auto oldType = operand.getType();
+  auto newType = typeConverter->convertType(oldType);
+  if (oldType == newType) {
+    return operand;
+  }
+  auto constOp = dyn_cast<arith::ConstantOp>(operand.getDefiningOp());
+  if (!constOp) {
+    return operand;
+  }
+  auto intAttr = constOp->getAttr("value").dyn_cast<IntegerAttr>();
+  APInt value = intAttr.getValue();
+  APInt newValue = value.truncSSat(newType.getIntOrFloatBitWidth());
+  if (newValue == value) {
+    return operand;
+  }
+  Value truncatedConst = rewriter.create<arith::ConstantOp>(
+      constOp->getLoc(), IntegerAttr::get(oldType, newValue));
+  return truncatedConst;
+}
+
+// Utility function to check that op is signed or not.
+bool isSignedOp(Operation *op) {
```

Comment:
This is fine for now I guess. But I'd think we need to define op traits in the upstream so that we don't need to maintain a list of ops here specifically by ourselves. That is, extend the [definitions in SPIR-V dialect](https://github.com/llvm/llvm-project/blob/main/mlir/include/mlir/Dialect/SPIRV/IR/SPIRVOpTraits.h#L23-L27) to all dialects. Let me create a patch for that.

Reply:
Sounds good, added this note!